src/dist/replication/lib/duplication/duplication_sync_timer.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2017-present, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <atomic>
+
+#include "dist/replication/lib/replica_stub.h"
+
+#include <dsn/dist/replication/duplication_common.h>
+#include <dsn/utility/chrono_literals.h>
+
+namespace dsn {
+namespace replication {
+
+using namespace literals::chrono_literals;
+
+constexpr int DUPLICATION_SYNC_PERIOD_SECOND = 10;
+
+// Per-server(replica_stub)-instance.
+class duplication_sync_timer
+{
+public:
+    explicit duplication_sync_timer(replica_stub *stub);
+
+    ~duplication_sync_timer();
+
+    void start();
+
+    void close();
+
+private:
+    // replica server periodically uploads current confirm points to meta server by sending
+    // `duplication_sync_request`.
+    // if success, meta server will respond with `duplication_sync_response`, which contains
+    // the entire set of duplications on this server.
+    void run();
+
+    /// \param dup_map: <appid -> list<dup_entry>>
+    void
+    update_duplication_map(const std::map<app_id, std::map<dupid_t, duplication_entry>> &dup_map);
+
+    void on_duplication_sync_reply(error_code err, const duplication_sync_response &resp);
+
+    std::vector<replica_ptr> get_all_primaries();
+
+    std::vector<replica_ptr> get_all_replicas();
```

Comment:
What the usage of 'get_all_primaries' and 'get_all_ replicas'?

Reply:
- `get_all_ replicas`: see `duplication_sync_timer::update_duplication_map`
- `get_all_primaries`: see `duplication_sync_timer::run`